### PR TITLE
Gatk4 combinegvcfs

### LINKED ISSUE #1341: new module: GATK4/COMBINEGVCFS
I think it would be good to have a module for [gatk4/combinegvcfs](https://bioconda.github.io/recipes/gatk4/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by sateeshperi**
## PR checklist

Closes #1341  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/gatk4/combinegvcfs/main.nf [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.4.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/gatk4:4.2.4.1--hdfd78af_0' :
+        'quay.io/biocontainers/gatk4:4.2.4.1--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(fasta), path(fastaIndex), path (fastaDict)
+    path(vcffiles)
+    path(vcf_idx)
```
Comment (maxulysse) [inline modules/gatk4/combinegvcfs/main.nf:13]: Shouldn't it be more something like that?
```suggestion
    tuple val(meta), path(fasta), path(vcf), path(vcf_idx)
    path (fastaIndex)
    path (fastaDict)
```

File: modules/gatk4/combinegvcfs/main.nf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def avail_mem       = 3
+    if (!task.memory) {
+        log.info '[GATK COMBINEGVCFS] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = task.memory.giga
+    }
+    def variant_string = ''
+    vcffiles.each {variant_string += "-V ${it} "} // loop to create a string adding -V to each vcf file
```
Comment (maxulysse) [inline modules/gatk4/combinegvcfs/main.nf:32]: def input_files = vcf.collect{"--V ${it}"}.join(' ')

File: modules/gatk4/combinegvcfs/main.nf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    gatk \\
+        --java-options "-Xmx${avail_mem}g" \\
+        CombineGVCFs \\
+        -R ${fasta} \\
+        -O ${prefix}.combined.g.vcf.gz \\
+        ${args} \\
+        --tmp-dir . \\
+        ${variant_string}
```
[the remark this answers is not in the document]
Comment (maxulysse) [reply modules/gatk4/combinegvcfs/main.nf:41]: @FriederikeHanssen I don't remember what we decided with the tmp dir, can you help me on this one?

File: modules/gatk4/combinegvcfs/meta.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      pattern: "*.fasta"
+  - fai:
+      type: file
+      description: FASTA index file
+      pattern: "*.{fai}"
+  - dict:
+      type: file
+      description: FASTA dictionary file
+      pattern: "*.{dict}"
+  - vcffiles:
```
Comment (maxulysse) [inline modules/gatk4/combinegvcfs/meta.yml:33]:
```suggestion
  - vcf:
```

File: modules/gatk4/combinegvcfs/main.nf [hunk 5]
```diff
@@ -0,0 +1,48 @@
+process GATK4_COMBINEGVCFS {
+    tag "$meta.id"
+    label 'process_low'
+
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.4.1" : null)
```
Comment (FriederikeHanssen) [inline modules/gatk4/combinegvcfs/main.nf:5]: would recommend using the newest gatk version here

File: modules/gatk4/combinegvcfs/main.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (FriederikeHanssen) [reply modules/gatk4/combinegvcfs/main.nf:41]: hm yes this looks right

[the remark this answers is not in the document]
Comment (FriederikeHanssen) [reply modules/gatk4/combinegvcfs/main.nf:41]: we could in theory though leave it in args, which I think we did for the other gatk4/modules. It is not a file and therefore per definition should be set in the modules.config

File: modules/gatk4/combinegvcfs/main.nf (hunk 1 above)
Comment (sateeshperi) [reply modules/gatk4/combinegvcfs/main.nf:13]: the vcf and vcf_idx are lists, not single instances and will be gathered via collect by the user. Would prefer the current input order

Comment (FriederikeHanssen) [reply modules/gatk4/combinegvcfs/main.nf:13]: list of files are still path objects. The fasta does not depend on the meta map though, does it? I mean it is one reference for all samples, but the list of vcf files belongs to the meta map and is used to gather the right vcfs files

Comment (mciprianoCDC) [reply modules/gatk4/combinegvcfs/main.nf:13]: @FriederikeHanssen -  You are correct in tis instance except the list of vcf files will belong to multiple meta maps. How should we deal with this?

Comment (maxulysse) [reply modules/gatk4/combinegvcfs/main.nf:13]: Remap the vcf files before this process and give this set a new meta map with an id that will be used as the output name

Comment (FriederikeHanssen) [reply modules/gatk4/combinegvcfs/main.nf:13]: the meta map I guess has a field for the splitting this should be removed. You will need to touch it anyways to group the right vcf files together for the input in the current implementation

Comment (mciprianoCDC) [reply modules/gatk4/combinegvcfs/main.nf:13]: So the input will be [meta, [list of vcf files], [list of vcf indexes] ], correct?

Comment (maxulysse) [reply modules/gatk4/combinegvcfs/main.nf:13]: Just meta, vcf, vcf_index is fine, as we're doing in the manta germline module for example that can accommodate multiple inputs: https://github.com/nf-core/modules/blob/master/modules/manta/germline/main.nf#L11

Comment (mciprianoCDC) [reply modules/gatk4/combinegvcfs/main.nf:13]: Sorry, my question might not have been clear - I was just talking about within the workflow. I understand the rational of this approach - all good here.

File: modules/gatk4/combinegvcfs/main.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (mciprianoCDC) [reply modules/gatk4/combinegvcfs/main.nf:41]: A difference might be whether it is required or not. For this module, the tmp_dir argument is required and the process will fail if not given. Would it be appropriate to search the args variable to see if it is set, then set it to a default if it is not found?

[the remark this answers is not in the document]
Comment (mciprianoCDC) [reply modules/gatk4/combinegvcfs/main.nf:41]: Understood, we just want to follow best nf-core practices, so appreciate your recommendations.

**Review by maxulysse — APPROVED** [review]

**Pull request merged**